From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 18 Apr 2026 15:11:48 -0500
Subject: [PATCH 11/46] 
 0015-win32u-Track-swapchain-reference-of-client-surface.patch

---
 dlls/win32u/vulkan.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/win32u/vulkan.c b/dlls/win32u/vulkan.c
index 310ce8923c9..b2307b873d8 100644
--- a/dlls/win32u/vulkan.c
+++ b/dlls/win32u/vulkan.c
@@ -141,6 +141,7 @@ struct surface
 {
     struct vulkan_surface obj;
     struct client_surface *client;
+    struct swapchain *swapchain;
     HWND hwnd;
 };
 
@@ -1867,6 +1868,8 @@ static VkResult win32u_vkCreateSwapchainKHR( VkDevice client_device, const VkSwa
 
     vulkan_object_init( &swapchain->obj.obj, host_swapchain );
     swapchain->surface = surface;
+    client_surface_add_ref(surface->client);
+    surface->swapchain = swapchain;
     swapchain->extents = create_info->imageExtent;
     instance->p_insert_object( instance, &swapchain->obj.obj );
 
@@ -1885,6 +1888,11 @@ void win32u_vkDestroySwapchainKHR( VkDevice client_device, VkSwapchainKHR client
     if (!swapchain) return;
 
     device->p_vkDestroySwapchainKHR( device->host.device, swapchain->obj.host.swapchain, NULL );
+    if (swapchain->surface)
+    {
+        swapchain->surface->swapchain = NULL;
+        client_surface_release( swapchain->surface->client );
+    }
     instance->p_remove_object( instance, &swapchain->obj.obj );
 
     free( swapchain );
